docker/docker_client.go

```diff
@@ -212,37 +208,48 @@ func hasFile(files []os.FileInfo, name string) bool {
 	return false
 }
 
-// newDockerClient returns a new dockerClient instance for refHostname (a host a specified in the Docker image reference, not canonicalized to dockerRegistry)
-// “write” specifies whether the client will be used for "write" access (in particular passed to lookaside.go:toplevelFromSection)
-func newDockerClient(ctx *types.SystemContext, ref dockerReference, write bool, actions string) (*dockerClient, error) {
+func newDockerClientFromRef(ctx *types.SystemContext, ref dockerReference, write bool, actions string) (*dockerClient, error) {
 	registry := reference.Domain(ref.ref)
-	if registry == dockerHostname {
-		registry = dockerRegistry
-	}
 	username, password, err := getAuth(ctx, reference.Domain(ref.ref))
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "error getting username and password")
 	}
+
+	return newDockerClientWithDetails(ctx, ref, registry, username, password, actions, write)
+}
+
+// newDockerClient returns a new dockerClient instance for refHostname (a host a specified in the Docker image reference, not canonicalized to dockerRegistry)
+// “write” specifies whether the client will be used for "write" access (in particular passed to lookaside.go:toplevelFromSection)
+func newDockerClientWithDetails(ctx *types.SystemContext, ref dockerReference, registry, username, password, actions string, write bool) (*dockerClient, error) {
 	tr := newTransport()
 	tr.TLSClientConfig = serverDefault()
+	client := &http.Client{Transport: tr}
+
 	// It is undefined whether the host[:port] string for dockerHostname should be dockerHostname or dockerRegistry,
 	// because docker/docker does not read the certs.d subdirectory at all in that case.  We use the user-visible
 	// dockerHostname here, because it is more symmetrical to read the configuration in that case as well, and because
 	// generally the UI hides the existence of the different dockerRegistry.  But note that this behavior is
 	// undocumented and may change if docker/docker changes.
-	certDir := dockerCertDir(ctx, reference.Domain(ref.ref))
-	if err := setupCertificates(certDir, tr.TLSClientConfig); err != nil {
-		return nil, err
+	var (
+		sigBase    signatureStorageBase
+		err        error
+		remoteName string
+	)
+	if ref != (dockerReference{}) {
```

Comment:
Like the previous `== "login"` conditional, this should instead be done by code structure, i.e. by moving the `configuredSignatureStorageBase` call to `newDockerClient`, and adding one more parameter to `newDockerClientWithDetails`.

Reply:
fixed